liberum_vault/src/vault.rs

```diff
@@ -0,0 +1,481 @@
+        once(0u64).chain(powers_from_1).last().unwrap()
+    }
+
+    fn power_2_desc_from_power_2(start: u64, limit: u64) -> u64 {
+        assert!(Self::is_power_of_2(start));
+
+        if limit == 0 {
+            return 0;
+        }
+
+        successors(Some(start), |&n| Some(n / 2))
+            .take_while(|x| x > &(limit / 2) || x == &limit)
+            .last()
+            .unwrap()
+    }
+
+    fn is_power_of_2(number: u64) -> bool {
+        number > 0 && (((number) & (number - 1)) == 0)
+    }
+
+    async fn prepare_db(&self) -> Result<()> {
+        const CREATE_FRAGMENT_TABLE_QUERY: &str = "
+            CREATE TABLE IF NOT EXISTS fragment (
+                hash0 INTEGER,
+                hash1 INTEGER,
+                hash2 INTEGER,
+                hash3 INTEGER,
+                path VARCHAR(255),
+                size INTEGER,
+                PRIMARY KEY (hash0, hash1, hash2, hash3)
+            )
+        ";
+
+        self.db
+            .call(|conn| Ok(conn.execute(CREATE_FRAGMENT_TABLE_QUERY, ())?))
+            .await?;
+
+        Ok(())
+    }
+
+    async fn load_fragment_info(&self, key: Key) -> Result<Option<FragmentInfo>> {
+        const SELECT_FRAGMENT_QUERY: &str = "
+            SELECT path, size
+            FROM fragment
+            WHERE hash0 = ?1 AND hash1 = ?2 AND hash2 = ?3 AND hash3 = ?4
+        ";
+
+        self.db
+            .call(move |conn| {
+                let mut stmt = conn.prepare(SELECT_FRAGMENT_QUERY)?;
+                let key_u64_slice = key.as_u64_slice_be();
+                let key_as_i64 = [
+                    key_u64_slice[0] as i64,
+                    key_u64_slice[1] as i64,
+                    key_u64_slice[2] as i64,
+                    key_u64_slice[3] as i64,
+                ];
+
+                let fragment = stmt
+                    .query_row(key_as_i64, |r| {
+                        let path: String = match r.get(0) {
+                            Ok(p) => p,
+                            Err(e) => return Result::Err(e),
+                        };
+
+                        let size: u64 = match r.get(1) {
+                            Ok(s) => s,
+                            Err(e) => return Result::Err(e),
+                        };
+
+                        Result::Ok(FragmentInfo::new(key, Path::new(&path), size))
+                    })
+                    .optional()?;
+
+                Ok(fragment)
+            })
+            .await
+            .map_err(|e| anyhow!(e))
+    }
+
+    async fn store_fragment_info(&self, fragment: FragmentInfo) -> Result<()> {
+        const SELECT_FRAGMENT_QUERY: &str = "SELECT COUNT(*) FROM fragment WHERE hash0 = ?1 AND hash1 = ?2 AND hash2 = ?3 AND hash3 = ?4";
+        const INSERT_FRAGMENT_QUERY: &str = "
+                INSERT INTO fragment (hash0, hash1, hash2, hash3, path, size)
+                VALUES (?1, ?2, ?3, ?4, ?5, ?6)
+            ";
+
+        let hash_as_u64 = fragment.hash.as_u64_slice_be();
```

Comment:
big endian?

Reply:
Duplicate comment, lets discuss on the previous one and use this one just to point to another occurence of this method